### PR TITLE
Test gen

### PR DESCRIPTION
**Pull request by noahdietz**

### CODE REVIEW

**Comment by noahdietz** [comment]
Still need to delete hard coded testing data and console logs after adding output formatting

**Comment by mohsen1** [comment]
Looks good to me from a quick glance. Please rebase from master

File: .eslintrc [hunk 1]
```diff
@@ -6,6 +6,7 @@
 ---
   env:
     node: true
+    mocha: true
```
Comment (mohsen1) [inline .eslintrc]: remove this.

File: index.js [hunk 2]
```diff
@@ -1 +1,281 @@
...
+ * @returns {string} generated test for response type
+ */
+function testGenResponse(swagger, path, operation, response, config) {
+  var result, gen, source, param,
+    // request payload
+    data = {
+      'responseCode': response,
+      'description': swagger.paths[path][operation].responses[response].description,
+      'assertion': config.assertionFormat,
+      'asynchronous': true
```
Comment (mohsen1) [inline index.js]: should go away

File: index.js [hunk 3]
```diff
@@ -1 +1,281 @@
...
+    };
+
+  // adding body parameters to payload
+  if (swagger.paths[path][operation].hasOwnProperty('parameters')) {
+    data.parameters = [];
+
+    // only adds body parameters to request, ignores query params
+    for (param in swagger.paths[path][operation].parameters) {
+      if (swagger.paths[path][operation].parameters[param].in === 'body') {
```
Comment (mohsen1) [inline index.js]: we need to support `formData` too

File: index.js [hunk 4]
```diff
@@ -1 +1,281 @@
...
+        data.parameters.push(swagger.paths[path][operation].parameters[param]);
+      }
+    }
+  }
+
+  // request url vs. supertest path
+  if (config.testmodule === 'request') {
+    data.url = swagger.schemes[0] + '://' + swagger.host +
```
Comment (mohsen1) [inline index.js]: There is no guarantee that `swagger.schemes[0]` exists

Same for `host`. Default to `localhost`

File: index.js [hunk 5]
```diff
@@ -1 +1,281 @@
...
+      (swagger.basePath !== undefined ? swagger.basePath : '') + path;
+  } else {
+    data.path = (swagger.basePath !== undefined ? swagger.basePath : '') + path;
```
Comment (mohsen1) [inline index.js]: DRY 50

File: index.js [hunk 6]
```diff
@@ -1 +1,281 @@
...
+  }
+
+  // template source decision logic
+  if (operation === 'get') {
```
Comment (mohsen1) [inline index.js]: Consider #7 for the way you organize your templates

File: index.js [hunk 7]
```diff
@@ -1 +1,281 @@
...
+    if (!data.hasOwnProperty('parameters') || data.parameters.length === 0) {
```
Comment (mohsen1) [inline index.js]: `if (!data.parameters.length)`

File: index.js [hunk 8]
```diff
@@ -1 +1,281 @@
...
+      source = read('./templates/' + config.testmodule
+        + '/get/get.handlebars', 'utf8');
```
Comment (mohsen1) [inline index.js]: use path module:

```js
  source = read(path.join('./templates/', config.testmodule, '/get/get.handlebars'), 'utf8')
```

File: index.js [hunk 9]
```diff
@@ -1 +1,281 @@
...
+  } else if (operation === 'post') {
+    source = read('./templates/' + config.testmodule
+      + '/post/post.handlebars', 'utf8');
+  } else if (operation === 'put') {
+    source = read('./templates/' + config.testmodule
+      + '/put/put/handlebars', 'utf8');
+  }
+
+  // compile template source and return test string
+  gen = handlebars.compile(source);
```
Comment (mohsen1) [inline index.js]: use a name like `templateFn` or something like that

File: index.js [hunk 10]
```diff
@@ -1 +1,281 @@
...
+    if (responses.hasOwnProperty(res)) {
+      result.push(testGenResponse(swagger,
+        path,
+        operation,
+        res,
+        config));
+    }
+  }
+
+  var output = "  describe('" + operation + "', function() {\n";
```
Comment (mohsen1) [inline index.js]: Lets use templates for:

* describe blocks 
* iteration

File: index.js [hunk 11]
```diff
@@ -1 +1,281 @@
...
+  var operations = swagger.paths[path],
+    result = [], op;
+
+  for (op in operations) {
+    if (operations.hasOwnProperty(op)) {
+      result.push(testGenOperation(swagger, path, op, config));
+    }
+  }
+
+  var output = "describe('" + path + "', function() {\n", test;
```
Comment (mohsen1) [inline index.js]: https://github.com/apigee-127/swagger-test-templates/pull/3/files#r33402249

File: index.js [hunk 12]
```diff
@@ -1 +1,281 @@
...
+    targets = config.pathNames,
+		result = [],
+    output = [],
+    path, ndx, test, i = 0;
+
+  var imports = "'use strict';\n\n"
+    + "var chai = require('chai'),\n"
+    + "  request = require('request'),\n"
+    + "  supertest = require('supertest');\n"
+    + "var api = supertest('" + swagger.schemes[0] + '://' + swagger.host + "'); // supertest init\n\n";
```
Comment (mohsen1) [inline index.js]: Be carful about:  `swagger.schemes[0] + '://' + swagger.host`

File: index.js [hunk 13]
```diff
@@ -1 +1,281 @@
...
+  if (config.pathNames.length === 0) {
+    // builds tests for all paths in API
+    for (path in paths) {
+      if (paths.hasOwnProperty(path)) {
+        result.push(testGenPath(swagger, path, config));
+      }
+    }
+  } else {
+      // loops over specified paths from config
+      for (path in targets) {
```
Comment (mohsen1) [inline index.js]: Indentation issue. Review ESLint rules, this should've been failed

File: index.js [hunk 14]
```diff
@@ -1 +1,281 @@
...
+        }
+      }
+  }
+
+
+  if (config.pathNames.length === 0) {
+    for (ndx in result) {
+      if (result.hasOwnProperty(ndx)) {
+        output.push({
+        'name': 'Stub.js',
```
Comment (mohsen1) [inline index.js]: name property name should not be wrapped in quotes. ESLint issue

File: index.js [hunk 15]
```diff
@@ -1 +1,281 @@
...
+        'test': imports + result[ndx]
+        });
+      }
+    }
+
+    // build file names from paths
+
+    for (path in paths) {
```
Comment (mohsen1) [inline index.js]: File writing logic should go away

File: index.js [hunk 16]
```diff
@@ -1 +1,281 @@
...
+
+  return output;
+}
+
+module.exports = {
+  testGen: testGen
+};
+
+// https://gist.github.com/doginthehat/1890659
+handlebars.registerHelper('compare', function(lvalue, rvalue, options) {
```
Comment (mohsen1) [inline index.js]: Lets name the helper `'is'`

File: package.json [hunk 17]
```diff
@@ -38,5 +39,11 @@
   "devDependencies": {
     "mocha": "^2.2.5",
     "mocha-eslint": "^0.1.7"
+  },
+  "dependencies": {
+    "chai": "^3.0.0",
```
Comment (mohsen1) [inline package.json]: Move testing dependencies to devDependencies

File: templates/request/get/get.handlebars [hunk 18]
```diff
@@ -0,0 +1,19 @@
+    it('should be {{description}}', function({{#if asynchronous}}done{{/if}}) {
```
Comment (mohsen1) [inline templates/request/get/get.handlebars]: Lets get rid of `asynchronous` option

Comment (mohsen1) [inline templates/request/get/get.handlebars]: We should have response code here.

File: templates/request/get/get.handlebars [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+      request({
+        url: '{{url}}',
+        method: 'GET',
+        headers: {'Content-Type': 'application/json'}
+      },
+      function(error, response, body) {
+        {{#compare assertion 'expect'}}
+        chai.expect(body).to.have.property('name');
```
Comment (mohsen1) [inline templates/request/get/get.handlebars]: Lets add some comments here about the generated code. It should be clear that it's dummy code

File: index.js (hunk 12 above)
Comment (mohsen1) [inline index.js]: requiring supertest and using it should be conditional.

And invoke `chai.should()`

File: test/.eslintrc [hunk 20]
```diff
@@ -2,3 +2,6 @@
   extends: ../.eslintrc
   env:
     mocha: true
+
+  rules:
+  	no-unused-vars: [2, {"var":"local", "args":"none"}]
```
Comment (mohsen1) [inline test/.eslintrc]: This is extra. See https://github.com/apigee-127/swagger-test-templates/pull/3/files#r33403397

File: test/minimal/swagger.json [hunk 21]
```diff
@@ -0,0 +1,106 @@
+{
```
Comment (mohsen1) [inline test/minimal/swagger.json]: This is not minimal anymore. Lets break it down.

File: test/minimal/test.js [hunk 22]
```diff
@@ -0,0 +1,87 @@
+var assert = require('chai').assert;
+var testGen = require('../../index.js').testGen;
+var swagger = require('./swagger.json');
+var lint = require('mocha-eslint');
+
+describe('minimal swagger', function(){
+	describe('request-option', function(){
+		describe('pathNames-option', function(){
```
Comment (mohsen1) [inline test/minimal/test.js]: Indentation errors here too. Lets fix the ESLint config

File: test/minimal/test.js [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+			it ('should still generate all paths from empty pathNames option', function(done){
```
Comment (mohsen1) [inline test/minimal/test.js]: If test is sync, don't include `done` argument

File: test/minimal/test.js [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+        var output = testGen(swagger, {
+          'assertionFormat':'should',
+          'pathNames':[],
+          'testmodule':'request'
+           ,'destination':'./test/minimal'
+        });
+
+        assert.isArray(output);
```
Comment (mohsen1) [inline test/minimal/test.js]: the describe block should say: `"testing with should and request"`:

almost each assertion should be wrapped in a `it` block.

**Pull request merged**